awscli/customizations/history/show.py

```diff
@@ -0,0 +1,463 @@
+        super(DetailedFormatter, self).__init__(output, include, exclude)
+        self._request_id_to_api_num = {}
+        self._num_api_calls = 0
+        self._colorize = colorize
+        if self._colorize:
+            colorama.init(autoreset=True, strip=False)
+
+    def _display(self, event_record):
+        section_definition = self._SECTIONS.get(event_record['event_type'])
+        if section_definition:
+            self._display_section(event_record, section_definition)
+
+    def _display_section(self, event_record, section_definition):
+        if 'title' in section_definition:
+            self._display_title(section_definition['title'], event_record)
+        for value_definition in section_definition['values']:
+            self._display_value(value_definition, event_record)
+
+    def _display_title(self, title, event_record):
+        formatted_title = self._format_section_title(title, event_record)
+        self._write_output(formatted_title)
+
+    def _display_value(self, value_definition, event_record):
+        value_description = value_definition['description']
+        event_record_payload = event_record['payload']
+        value = event_record_payload
+        if 'payload_key' in value_definition:
+            value = event_record_payload[value_definition['payload_key']]
+        formatted_value = self._format_description(value_description)
+        formatted_value += self._format_value(
+            value, event_record, value_definition.get('value_format')
+        )
+        self._write_output(formatted_value)
+
+    def _write_output(self, content):
+        if isinstance(content, six.text_type):
+            content = content.encode('utf-8')
+        self._output.write(content)
+
+    def _format_section_title(self, title, event_record):
+        formatted_title = title
+        api_num = self._get_api_num(event_record)
+        if api_num is not None:
+            formatted_title = ('[%s] ' % api_num) + formatted_title
+        formatted_title = self._color_if_configured(formatted_title, 'title')
+        formatted_title += '\n'
+
+        formatted_timestamp = self._format_description('at time')
+        formatted_timestamp += self._format_value(
+            event_record['timestamp'], event_record, value_format='timestamp')
+
+        return '\n' + formatted_title + formatted_timestamp
+
+    def _get_api_num(self, event_record):
+        request_id = event_record['request_id']
+        if request_id:
+            if request_id not in self._request_id_to_api_num:
+                self._request_id_to_api_num[
+                    request_id] = self._num_api_calls
+                self._num_api_calls += 1
+            return self._request_id_to_api_num[request_id]
+
+    def _format_description(self, value_description):
+        return self._color_if_configured(
+            value_description + ': ', 'description')
+
+    def _format_value(self, value, event_record, value_format=None):
```

Comment:
This class is really hard to follow.  You have two levels of dispatch here:

* `_display -> _display_*`: no dynamic dispatch, calls a series of `_format_*` calls
* `_format_value -> _format_*`: dynamic dispatch

What I'd suggest for this PR:

* If you're going to use dynamic dispatch in this class, you need to make sure you use a unique prefix that cannot interfere with other methods in this class (consider if you called `_format_value` with `value_format='value'`).
* Use an entirely different prefix for things that use dynamic dispatch.  It's confusing that some `_format_*` calls are called by `_display_*` and some are called by `_format_value`.

Going forward in future code, some suggestions to consider:

* Consider separate classes for your dynamic dispatch.  The only other methods in the class should be methods used by the methods you dispatch to.
* Have the dynamic dispatch map to some public method (e.g `serialize() -> _serialize_{type_name}`, `visit() -> _visit_{node_type}`, `parse() -> _parse_{type}`, `format() -> _format_{name}`)

Reply:
Makes sense. To fix it, I decided to rip out the value specific format dispatchers all to its own class that the `DetailedFormatter` uses to pretty format the values for each section. Hopefully that will be clearer.